src/MongoDB.Driver.Core/Core/Clusters/IClusterExtensions.cs

```diff
@@ -0,0 +1,79 @@
+/* Copyright 2019–present MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System.Threading;
+using System.Threading.Tasks;
+using MongoDB.Driver.Core.Bindings;
+using MongoDB.Driver.Core.Clusters.ServerSelectors;
+using MongoDB.Driver.Core.Servers;
+
+namespace MongoDB.Driver.Core.Clusters
+{
+    /// <summary>
+    /// This class contains extension methods for ICluster used for server selection with sharded transactions.
+    /// </summary>
+    internal static class IClusterExtensions
+    {
+        public static IServer SelectServerAndPinIfNeeded(
+            this ICluster cluster,
+            ICoreSessionHandle session,
+            IServerSelector selector,
+            CancellationToken cancellationToken)
+        {
+            if (IsInShardedTransaction(cluster, session))
+            {
+                if (session.PinnedShardRouter == null)
+                {
+                    session.PinnedShardRouter = cluster.SelectServer(selector, cancellationToken);
+                }
+
+                return session.PinnedShardRouter;
+            }
+            else
+            {
+                return cluster.SelectServer(selector, cancellationToken);
+                
+            }
+        }
+        
+        public static async Task<IServer> SelectServerAndPinIfNeededAsync(
+            this ICluster cluster,
+            ICoreSessionHandle session,
+            IServerSelector selector,
+            CancellationToken cancellationToken)
+        {
+            if (IsInShardedTransaction(cluster, session))
+            {
+                if (session.PinnedShardRouter == null)
+                {
+                    session.PinnedShardRouter =
+                        await cluster.SelectServerAsync(selector, cancellationToken).ConfigureAwait(false);
+                }
+
+                return session.PinnedShardRouter;
+            }
+            else
+            {
+                return await cluster.SelectServerAsync(selector, cancellationToken).ConfigureAwait(false);   
+            }
+            
+        }
+
+        private static bool IsInShardedTransaction(ICluster cluster, ICoreSessionHandle session)
+        {
+            return session != null && session.IsInTransaction && cluster.Description.Type == ClusterType.Sharded;
```

Comment:
What happens if this transaction is the very first use of the cluster after construction.  Could the cluster type be Unknown, and you would therefore neglect to pin for the very first transaction?

Reply:
That's a good point: I'm actually a little unsure. @rstam , what do you think?

Reply:
That sounds like a race condition that can happen.

The first time we get here the cluster type might still be Unknown.

Reply:
Race condition resolved per offline discussion with Robert. I've added two extra comments and tweaked the method names. Local testing has seems to indicate that the unpinning of the ``mongos`` from while performing server selection does not appear to cause the test failures that Isabella saw when she unpinned when aborting a transaction, so I've left it in but will yank if it Evergreen complains later.